unix_integration/src/tool.rs

```diff
@@ -58,37 +61,99 @@ async fn main() -> ExitCode {
+                        ClientResponse::PamStatus(Some(false)) => {
+                            println!("auth failed!");
+                            break;
+                        }
+                        ClientResponse::PamStatus(None) => {
+                            println!("auth user unknown");
+                            break;
+                        }
+                        _ => {
+                            // unexpected response.
+                            error!("Error: unexpected response -> {:?}", r);
+                            break;
+                        }
+                    },
+                    Err(e) => {
+                        error!("Error -> {:?}", e);
+                        break;
                     }
-                    ClientResponse::PamStatus(Some(false)) => {
-                        println!("auth failed!");
+                }
+
+                match prompt.style {
+                    PamMessageStyle::PamPromptEchoOff => {
+                        let password = match rpassword::prompt_password(prompt.msg) {
+                            Ok(p) => p,
+                            Err(e) => {
+                                error!("Problem getting input: {}", e);
+                                return ExitCode::FAILURE;
+                            }
+                        };
+                        match prompt.cred_type {
+                            Some(CredType::Password) => {
+                                req = ClientRequest::PamAuthenticateStep(Some(PamCred::Password(
+                                    password,
+                                )));
+                            }
+                            _ => {
+                                req = ClientRequest::PamAuthenticateStep(Some(PamCred::MFACode(
+                                    password,
+                                )));
+                            }
+                        }
                     }
-                    ClientResponse::PamStatus(None) => {
-                        println!("auth user unknown");
+                    PamMessageStyle::PamPromptEchoOn => {
+                        let mut passcode = String::new();
+                        match io::stdin().read_line(&mut passcode) {
+                            Ok(_) => {}
+                            Err(e) => {
+                                error!("Problem getting input: {}", e);
+                                return ExitCode::FAILURE;
+                            }
+                        }
+                        passcode = passcode.trim_end_matches('\n').to_string();
+                        match prompt.cred_type {
+                            Some(CredType::Password) => {
+                                req = ClientRequest::PamAuthenticateStep(Some(PamCred::Password(
+                                    passcode,
+                                )));
+                            }
+                            _ => {
```

Comment:
You really should never get None here, since you're requesting a PamPromptEchoOn, but I handled it anyway as an MFA prompt. The point is just to prevent caching the response unless we are certain it is a password.
Same goes for line 117 (and the similar lines in the pam module).